src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/TagHelperSemanticSpanVisitor.cs

```diff
@@ -0,0 +1,217 @@
+            if (node.TagHelperAttributeInfo.Bound)
+            {
+                var result = CreateSemanticRange(node.Name, SyntaxKind.MarkupMinimizedTagHelperAttribute);
+                AddNode(result);
+            }
+
+            base.VisitMarkupMinimizedTagHelperAttribute(node);
+        }
+
+        public override void VisitMarkupTagHelperAttribute(MarkupTagHelperAttributeSyntax node)
+        {
+            if (node.TagHelperAttributeInfo.Bound)
+            {
+                var result = CreateSemanticRange(node.Name, SyntaxKind.MarkupTagHelperAttribute);
+                AddNode(result);
+            }
+
+            base.VisitMarkupTagHelperAttribute(node);
+        }
+
+        public override void VisitMarkupTagHelperDirectiveAttribute(MarkupTagHelperDirectiveAttributeSyntax node)
+        {
+            if (node.TagHelperAttributeInfo.Bound)
+            {
+                var transition = CreateSemanticRange(node.Transition, SyntaxKind.Transition);
+                AddNode(transition);
+
+                var directiveAttribute = CreateSemanticRange(node.Name, SyntaxKind.MarkupTagHelperDirectiveAttribute);
+                AddNode(directiveAttribute);
+
+                if (node.Colon != null)
+                {
+                    var colon = CreateSemanticRange(node.Colon, SyntaxKind.Colon);
+                    AddNode(colon);
+                }
+
+                if (node.ParameterName != null)
+                {
+                    var parameterName = CreateSemanticRange(node.ParameterName, SyntaxKind.MarkupTagHelperDirectiveAttribute);
+                    AddNode(parameterName);
+                }
+            }
+
+            base.VisitMarkupTagHelperDirectiveAttribute(node);
+        }
+
+        public override void VisitMarkupMinimizedTagHelperDirectiveAttribute(MarkupMinimizedTagHelperDirectiveAttributeSyntax node)
+        {
+
+            if (node.TagHelperAttributeInfo.Bound)
+            {
+                var transition = CreateSemanticRange(node.Transition, SyntaxKind.Transition);
+                AddNode(transition);
+
+                var directiveAttribute = CreateSemanticRange(node.Name, SyntaxKind.MarkupMinimizedTagHelperDirectiveAttribute);
+                AddNode(directiveAttribute);
+
+                if (node.Colon != null)
+                {
+                    var colon = CreateSemanticRange(node.Colon, SyntaxKind.Colon);
+                    AddNode(colon);
+                }
+
+                if (node.ParameterName != null)
+                {
+                    var parameterName = CreateSemanticRange(node.ParameterName, SyntaxKind.MarkupMinimizedTagHelperDirectiveAttribute);
+
+                    AddNode(parameterName);
+                }
+            }
+
+            base.VisitMarkupMinimizedTagHelperDirectiveAttribute(node);
+        }
+
+        private void AddNode(SemanticRange syntaxResult)
+        {
+            if (_range is null || syntaxResult.Range.OverlapsWith(_range))
```

Comment:
Should this also add a condition when `syntaxResult.Range` is null?

Reply:
That should not be possible. I've added a null-check to `SemanticRange` to reflect that.